File: h-index-ii.js
// Time Complexity : O(logN)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * @param {number[]} citations
 * @return {number}
 */
var hIndex = function(citations) {
    if (!citations || !citations.length) return 0;
    
    let left = 0,
        right = citations.length - 1;
    while (left <= right) {
        let mid = Math.floor((left + right) / 2);
        let currH = citations[mid];
        if (citations[mid] == citations.length - mid) return citations.length - mid;
        else if (citations[mid] < citations.length - mid) left = mid + 1;
        else right = mid - 1;
    }
    
    return citations.length - left;
};

File: h_IndexII.java

public class h_IndexII {

	public int hIndex(int[] citations) {
		
		if(citations == null || citations.length == 0)
			return 0;
		
		int left = 0, right = citations.length;
		int n = citations.length;
		
		while(left < right) {
			int mid = left + (right - left)/2;
			
			if(citations[mid] == n-mid)
				return n - mid;
			else if(citations[mid] < n - mid)
				left = mid+1;
			else
				right = mid;
		}
		return n - left;
	}
	
	public static void main(String[] args) {
		
		h_IndexII hi = new h_IndexII();
		int[] citations = {1, 2, 4, 7, 9, 10, 11};
		System.out.println(hi.hIndex(citations));
	}
}
